common/src/main/java/org/cloudfoundry/identity/uaa/authorization/UaaAuthorizationEndpoint.java

```diff
@@ -399,14 +409,16 @@ private String getSuccessfulRedirect(AuthorizationRequest authorizationRequest,
         }
 
         UriComponentsBuilder template = UriComponentsBuilder.fromUriString(authorizationRequest.getRedirectUri());
-        template.queryParam("code", authorizationCode);
+        //decode already encoded values
+        template = UriComponentsBuilder.fromHttpUrl(template.build(true).toUriString());
+        template.queryParam("code", encode(authorizationCode));
```

Comment:
Here we are proceeding to assume that the input URL is always encoded. Thus using the UriComponentsBuilder, we can only add already encoded parameters. Each parameter that is added, must be encoded before it is added.